### PR TITLE
implementation of the federated share provider

### PR DESCRIPTION
**Pull request by schiessle**
first iteration of the federated share provider. Still need to move over some stuff from the files_sharing app:
- All the mount/umount logic should be here. 
- Also the settings need to be moved. 

We can also move both parts in a second step. We also keep some stuff for local shares in the files_sharing app for now and clean this up in the next iteration.

### CODE REVIEW

**Comment by schiesbn** [comment]
This is all still work in progress... @rullzer please have a first look at the share provider if you have some time.

**Comment by schiesbn** [comment]
does anyone see why: ````PHP Fatal error:  Class 'OCA\FederatedFileSharing\TokenHandler' not found in /ssd/jenkins/workspace/core-ci-linux/database/sqlite/label/SLAVE/apps/federatedfilesharing/tests/lib/tokenHandlerTest.php on line 46````

**Comment by rullzer** [comment]
@schiesbn there are capital letters in your file names.

**Comment by rullzer** [comment]
Ok updated the PR. it is now working here. And looking pretty slick from my POV.

TODO:
- [ ] Tests!
- [ ] Handle disabled outgoing federated shares

**Comment by rullzer** [comment]
Test will come but please already have a look.

* The first commit is the actual app.
* Following commits are taking care of all the internal rewireing.

Right now the app is always enabled. At the moment I do not see a quick way around this. As the incoming federated stuff is not yet in this app. So hopefully we can clear this up soon.

CC: @DeepDiver1975 @PVince81 @schiesbn @nickvergessen

**Comment by PVince81** [comment]
- [ ] BUG: share folder with another user "user2@xxx". Accept the share. Then unshare from self. The owner still sees the federated share. (on stable8.2 it works, the share is gone)

**Comment by PVince81** [comment]
- [x] TEST: existing fed share appears in dialog
- [x] TEST: fed share over share dialog
- [x] TEST: shows error when sharing with unknown remote user
- [x] TEST: fed share with link share
- [ ] TEST: unshare from self :no_entry_sign: see above

File: apps/federatedfilesharing/lib/addresshandler.php [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+	 * @param string $user2
+	 * @param string $server2
+	 * @return bool true if both users and servers are the same
+	 */
+	public function compareAddresses($user1, $server1, $user2, $server2) {
+		$normalizedServer1 = strtolower($this->removeProtocolFromUrl($server1));
+		$normalizedServer2 = strtolower($this->removeProtocolFromUrl($server2));
+
+		$normalizedServer1 = $this->removeProtocolFromUrl($normalizedServer1);
+		$normalizedServer2 = $this->removeProtocolFromUrl($normalizedServer2);
```
Comment (PVince81) [inline apps/federatedfilesharing/lib/addresshandler.php]: twice ?

File: apps/federatedfilesharing/tests/lib/addressHandlerTest.php [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+namespace OCA\FederatedFileSharing\Tests\lib;
+
+
+use OCA\FederatedFileSharing\AddressHandler;
+use OCP\IL10N;
+use OCP\IURLGenerator;
+use Test\TestCase;
+
+class addressHandlerTest extends TestCase {
```
Comment (PVince81) [inline apps/federatedfilesharing/tests/lib/addressHandlerTest.php]: Upper case class name

File: apps/federatedfilesharing/tests/lib/tokenHandlerTest.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+
+namespace OCA\FederatedFileSharing\Tests\lib;
+
+
+use OCA\FederatedFileSharing\TokenHandler;
+use OCP\Security\ISecureRandom;
+use Test\TestCase;
+
+class tokenHandlerTest extends TestCase {
```
Comment (PVince81) [inline apps/federatedfilesharing/tests/lib/tokenHandlerTest.php]: uppercase class name

File: core/shipped.json [hunk 4]
```diff
@@ -8,6 +8,7 @@
     "enterprise_key",
     "external",
     "federation",
+    "federatedfilesharing",
```
Comment (PVince81) [inline core/shipped.json]: what's the difference with the "federation" app ?

File: lib/private/share20/providerfactory.php [hunk 5]
```diff
@@ -63,28 +69,90 @@ protected function defaultShareProvider() {
...
+			/*
+			 * Check if the app is enabled
+			 */
+			$appManager = $this->serverContainer->getAppManager();
+			if (!$appManager->isEnabledForUser('federatedfilesharing')) {
+				return null;
+			}
+
+			/*
+			 * TODO: add factory to federated sharing app
```
Comment (PVince81) [inline lib/private/share20/providerfactory.php]: remaining TODO, or for the future ?

File: lib/private/share20/providerfactory.php [hunk 6]
```diff
@@ -63,28 +69,90 @@ protected function defaultShareProvider() {
...
 	 * @inheritdoc
 	 */
 	public function getProvider($id) {
+		$provider = null;
 		if ($id === 'ocinternal') {
-			return $this->defaultShareProvider();
+			$provider = $this->defaultShareProvider();
+		}
+
+		if ($id === 'ocFederatedSharing') {
```
Comment (PVince81) [inline lib/private/share20/providerfactory.php]: this could be `else if`

**Comment by PVince81** [comment]
Code looks good otherwise.

File: apps/federatedfilesharing/appinfo/routes.php [hunk 7]
```diff
@@ -0,0 +1 @@
+<?php
```
Comment (nickvergessen) [inline apps/federatedfilesharing/appinfo/routes.php]: remove?

File: apps/federatedfilesharing/lib/addresshandler.php [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+	}
+
+	/**
+	 * split user and remote from federated cloud id
+	 *
+	 * @param string $address federated share address
+	 * @return array [user, remoteURL]
+	 * @throws HintException
+	 */
+	public function splitUserRemote($address) {
```
Comment (nickvergessen) [inline apps/federatedfilesharing/lib/addresshandler.php]: Can this method be OCP somewhere?

File: apps/federatedfilesharing/lib/addresshandler.php (hunk 1 above)
Comment (nickvergessen) [inline apps/federatedfilesharing/lib/addresshandler.php]: two servers two strips, sounds legit

Comment (PVince81) [inline apps/federatedfilesharing/lib/addresshandler.php]: I mean `remoteProtocolFromUrl` is called twice for the same server.

Comment (PVince81) [inline apps/federatedfilesharing/lib/addresshandler.php]: could be a copy-paste issue. On line 123 it's stripped with strtolower and then again here without strtolower.

Comment (rullzer) [inline apps/federatedfilesharing/lib/addresshandler.php]: mmm I have no idea... but I assume the second one is pretty useless

File: core/shipped.json (hunk 4 above)
Comment (rullzer) [inline core/shipped.json]: The federation app is syncing of address books if I'm not mistaken.

Anyway as I understand it we want to have a small app that we can actually fully uderstand and test.

File: lib/private/share20/providerfactory.php (hunk 5 above)
Comment (rullzer) [inline lib/private/share20/providerfactory.php]: future

File: apps/federatedfilesharing/lib/addresshandler.php (hunk 8 above)
Comment (rullzer) [inline apps/federatedfilesharing/lib/addresshandler.php]: Probably makes sense... but we can do that in a later PR... (beacuse... where to put it...)

Comment (nickvergessen) [inline apps/federatedfilesharing/lib/addresshandler.php]: `OC\Share\Helper::splitUserRemote()` duplicates?

**Comment by rullzer** [comment]
> BUG: share folder with another user "user2@xxx". Accept the share. Then unshare from self. The owner still sees the federated share. (on stable8.2 it works, the share is gone)

I think this is due to the share not actually being deleted from the source instance... it is just set to not accept... probabaly so we can later reaccept or something (kind of like with deleted group shares)...

Let me look into that

**Comment by rullzer** [comment]
@PVince81 mmm I can't see why it should not work as is...

Can you double check? Here the entry is properly deleted from the oc_shares table... and does not show up in the shared list...

**Comment by PVince81** [comment]
@rullzer actually I cheated and used a fed share to a local user. Should still work as it did in stable8.2.

Let me re-try.

**Comment by PVince81** [comment]
Okay, it works now. Weird. Maybe I tested it wrong or a rebase fixed it somehow.

:+1:

File: apps/federatedfilesharing/lib/addresshandler.php (hunk 8 above)
Comment (schiesbn) [inline apps/federatedfilesharing/lib/addresshandler.php]: `OC\Share\Helper::splitUserRemote()`

This method will no longer exists if the transition to Sharing 2.0 is finished. 99% of the code you see here already exists in OC\Share

File: core/shipped.json (hunk 4 above)
Comment (schiesbn) [inline core/shipped.json]: yes, federation app exchanges user lists between servers. The "federatedfilesharing" app is the federated file sharing provider for sharing 2.0

File: apps/federatedfilesharing/lib/notifications.php [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+	}
+
+	/**
+	 * try http post first with https and then with http as a fallback
+	 *
+	 * @param string $url
+	 * @param array $fields post parameters
+	 * @return array
+	 */
+	private function tryHttpPost($url, $fields) {
```
Comment (LukasReschke) [inline apps/federatedfilesharing/lib/notifications.php]: `array $fields`

File: apps/federatedfilesharing/lib/addresshandler.php [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+		if ($pos !== false) {
+			$user = substr($id, 0, $pos);
+			$remote = substr($id, $pos + 1);
+			$remote = $this->fixRemoteURL($remote);
+			if (!empty($user) && !empty($remote)) {
+				return array($user, $remote);
+			}
+		}
+
+		$l = \OC::$server->getL10N('core');
```
Comment (LukasReschke) [inline apps/federatedfilesharing/lib/addresshandler.php]: Any reason to not use `$this->l`?

File: apps/federatedfilesharing/lib/addresshandler.php [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+		$hint = $l->t('Invalid Federated Cloud ID');
+		throw new HintException('Invalid Fededrated Cloud ID', $hint);
```
Comment (LukasReschke) [inline apps/federatedfilesharing/lib/addresshandler.php]: Typo "Fededrated"

File: apps/federatedfilesharing/lib/federatedshareprovider.php [hunk 12]
```diff
@@ -0,0 +1,558 @@
...
+		list(, $remote) = $this->addressHandler->splitUserRemote($share->getSharedWith());
+		$this->notifications->sendRemoteUnShare($remote, $share->getId(), $share->getToken());
+	}
+
+	/**
+	 * @inheritdoc
+	 */
+	public function deleteFromSelf(IShare $share, $recipient) {
+		// nothing to do here. Technically deleteFromSelf in the context of federated
+		// shares is a umount of a external storage. This is handled here
```
Comment (LukasReschke) [inline apps/federatedfilesharing/lib/federatedshareprovider.php]: umount? Unmount?

File: apps/files_sharing/api/share20ocs.php [hunk 13]
```diff
@@ -195,7 +203,15 @@ public function deleteShare($id) {
...
 		if ($share === null) {
-			return \OCA\Files_Sharing\API\Local::deleteShare(['id' => $id]);
+			if (!$this->shareManager->outgoingServer2ServerSharesAllowed()) {
+				return new \OC_OCS_Result(null, 404, 'wrong share ID, share doesn\'t exist.');
+			}
+
+			try {
+				$share = $this->shareManager->getShareById('ocFederatedSharing:' . $id);
+			} catch (ShareNotFound $e) {
+				return new \OC_OCS_Result(null, 404, 'wrong share ID, share doesn\'t exist.');
```
Comment (LukasReschke) [inline apps/files_sharing/api/share20ocs.php]: Move error message into variable / static?

File: apps/federatedfilesharing/lib/federatedshareprovider.php [hunk 14]
```diff
@@ -0,0 +1,558 @@
...
+		$qb->andWhere($qb->expr()->eq('share_type', $qb->createNamedParameter(self::SHARE_TYPE_REMOTE)));
+
+		/**
+		 * Reshares for this user are shares where they are the owner.
+		 */
+		if ($reshares === false) {
+			//Special case for old shares created via the web UI
+			$or1 = $qb->expr()->andX(
+				$qb->expr()->eq('uid_owner', $qb->createNamedParameter($userId)),
+				$qb->expr()->isNull('uid_initiator')
```
Comment (LukasReschke) [inline apps/federatedfilesharing/lib/federatedshareprovider.php]: Ah uses func_get_args … grrrr… Let's fix the PHPDocs of the method then?

Comment (LukasReschke) [inline apps/federatedfilesharing/lib/federatedshareprovider.php]: Awesome. The notation for that comes only with PHP 5.6 let's ignore that then for now :speak_no_evil:

**Comment by schiesbn** [comment]
Fixed some small doc-bugs and issues reported by @LukasReschke... Will also take care of the other issues reported by Lukas.

@rullzer Thanks a lot for moving this PR forward!

**Comment by LukasReschke** [comment]
Besides my nitpicking: :+1:

File: apps/federatedfilesharing/lib/federatedshareprovider.php (hunk 14 above)
Comment (schiesbn) [inline apps/federatedfilesharing/lib/federatedshareprovider.php]: Not that easy... Looked at it with @nickvergessen some time ago. It's more a issue of PHPStorm who can't interpret the doc correctly... At least that was the case when we looked at it a few weeks ago.

Anyway... Out of scope for this PR

File: apps/federatedfilesharing/lib/addresshandler.php (hunk 10 above)
Comment (schiesbn) [inline apps/federatedfilesharing/lib/addresshandler.php]: no, just a left over from copy&paste... fixed

File: apps/federatedfilesharing/lib/federatedshareprovider.php [hunk 15]
```diff
@@ -0,0 +1,558 @@
...
+	 *
+	 * @param int $itemSource
+	 * @param string $shareWith
+	 * @param string $sharedBy
+	 * @param string $uidOwner
+	 * @param int $permissions
+	 * @param string $token
+	 * @return int
+	 */
+	private function addShareToDB($itemSource, $itemType, $shareWith, $sharedBy, $uidOwner, $permissions, $token) {
```
Comment (schiesbn) [inline apps/federatedfilesharing/lib/federatedshareprovider.php]: same as above... phpstorm issue and out of scope for this PR

File: apps/federatedfilesharing/lib/federatedshareprovider.php (hunk 12 above)
Comment (schiesbn) [inline apps/federatedfilesharing/lib/federatedshareprovider.php]: every Linux user know that it is called umount :wink:

File: apps/federatedfilesharing/lib/federatedshareprovider.php (hunk 15 above)
Comment (LukasReschke) [inline apps/federatedfilesharing/lib/federatedshareprovider.php]: $itemType is clearly missing in the DocBlock?

Comment (schiesbn) [inline apps/federatedfilesharing/lib/federatedshareprovider.php]: already added

**Comment by schiesbn** [comment]
code added by @rullzer looks good... Thanks! :+1:

**Pull request merged**